File: ygg.py
from http.client import HTTPResponse
from urllib.error import URLError
import urllib.request as url

import re
import json
from typing import NamedTuple
from enum import IntFlag
from pathlib import Path


class OpenType(IntFlag):
    Closed = 0
    Left   = 1
    Right  = 2
    Both   = 3


class Range:
    def __init__(self, min_num: int, max_num: int, open_type: OpenType = OpenType.Closed):
        self.min : int      = min_num
        self.max : int      = max_num
        self.open: OpenType = open_type

    def __contains__(self, other: int | Range):
        if self.open == OpenType.Both:
            return True

        if isinstance(other, Range):
            left, right = other.min, other.max
        else:
            left, right = other, other

        return (OpenType.Left in self.open or self.min <= left)\
            and (OpenType.Right in self.open or self.max >= right)

    def __str__(self) -> str:
        return f"{self.min}-{self.max}"

    def to_json(self) -> dict[str, int]:
        return { "min": self.min, "max": self.max, "open": self.open}

    @classmethod
    def from_str(cls, s: str):
        s = s.strip()

        o = OpenType.Closed
        if s and s[-1] == '+':
            o = OpenType.Right

        min_num, max_num = 1, 1
        if (nums := sorted([int(i[0]) for i in re.finditer(r"\d+", s)])):
            min_num, max_num = nums[0], nums[-1]

        return cls(min_num, max_num, o)


class Entry(NamedTuple):
    id     : str
    name   : str
    lang   : str
    players: Range


def read_site() -> str:
    res = None
    try:
        res = url.urlopen("https://yggdrasil.cafe/pages/spieleauswahl")
        if isinstance(res, HTTPResponse):
            return res.read().decode("utf-8")
    except URLError:
        pass
    return ""


ENTRY_PATTERN = re.compile(
    r"<tr[^>]+>\s*"
    r"<td[^>]+>(?P<id>[^<]+).*?</td>\s*"
    r"<td[^>]+>(?P<name>[^<]+).*?</td>\s*"
    r"<td[^>]+>(?P<lang>[^<]+).*?</td>\s*"
    r"<td[^>]+>(?P<cnt>[^<]+).*?</td>\s*"
    r"</tr>", re.DOTALL
)


def json_encode(obj):
    if isinstance(obj, Range):
        return obj.to_json()
    raise TypeError("Cannot serialize")


if __name__ == "__main__":
    html = read_site()
    table = re.search("<table[^>]+(.*)</table>", html, re.DOTALL)

    if table is None:
        print("no table found")
        exit(0)

    db = [Entry(
        tr["id"],
        tr["name"],
        tr["lang"],
        Range.from_str(tr["cnt"]))
        for tr in ENTRY_PATTERN.finditer(table[0])]

    with Path("ygg_db.json").open("w+") as fl:
        json.dump(db, fl, default=json_encode)

File: ygg.ts
enum OpenType {
    Closed = 0,
    Left   = 1,
    Right  = 2,
    Both   = 3
}


class OpenRange {
    constructor(
	public min: number,
	public max: number,
	public open: OpenType = OpenType.Closed) {}

    contains(other: number | OpenRange): boolean {
	const isRange = other instanceof OpenRange;
	const left    = isRange ? other.min : other;
	const right   = isRange ? other.max : other;

	const leftValid = (this.open & OpenType.Left) !== 0 || this.min <= left;
	const rightValid = (this.open & OpenType.Right) !== 0 || this.max >= right;

	return leftValid && rightValid;
    }

    toString(): string {
	return `${this.min} - ${this.max}${(this.open & OpenType.Right) !== 0 ? '+' : ''}`;
    }

    static fromString(s: string): OpenRange {
	let st = s.trim();
	let open = st.at(-1) == '+' ? OpenType.Right : OpenType.Closed;
	let nums = (st.match(/\d+/g) || []).map(Number).sort();

	if (nums)
	    return new this(nums[0], nums.at(-1) || nums[0], open);
	return new this(1, 1, open);
    }
}


interface GameEntry {
    id: string;
    name: string;
    lang: string;
    players: OpenRange
}


interface GameEntryDTO extends Omit<GameEntry, "players"> {
    players: {
	min: number;
	max: number;
	open: number;
    };
}


let gameList: GameEntry[] = []

function displayGameList(data: GameEntry[]) {
    const ul = document.getElementById("list");

    if (!ul)
	return;

    ul.replaceChildren()

    const fragment = document.createDocumentFragment();

    data.forEach(d => {
	const li = document.createElement("li");

	const idSpan = document.createElement("span");
	idSpan.className = "game_id";
	idSpan.textContent = d.id;

	const nameSpan = document.createElement("span");
	nameSpan.className = "game_name";
	nameSpan.textContent = d.name;

	const playersSpan = document.createElement("span");
	playersSpan.className = "game_players";
	playersSpan.textContent = d.players.toString();

	li.append(idSpan, nameSpan, playersSpan);
	fragment.appendChild(li);
    });

    ul.appendChild(fragment);
}


async function loadGameList(): Promise<GameEntry[]> {
    try {
	const response = await fetch("./ygg_db.json");
	if (!response.ok)
	    throw new Error(`HTTP Error! ${response.statusText}`);

	const data: GameEntryDTO[] = await response.json();

	return data.map(entry => ({
	    ...entry,
	    players: new OpenRange(entry.players.min, entry.players.max, entry.players.open)
	}));
    }
    catch (err) {
	console.error(err);
	return [];
    }
}

const filterInput = document.getElementById("filter") as HTMLInputElement | null;

filterInput?.addEventListener("input", event => {
    const target = event.target as HTMLInputElement;
    let value = target.value.toLowerCase();

    if (value == "") {
	displayGameList(gameList);
	return;
    }

    const playersPattern = /\s*p:\s*([\d\-\s\+]+)/i
    const players = value.match(playersPattern)

    let displayList = gameList;

    if (players) {
	const range = OpenRange.fromString(players[1]);
	console.log(range.toString());
	displayList = displayList.filter(e => e.players.contains(range))
	value = value.replace(playersPattern, "")
    }

    // const re = new RegExp(value, "i");
    displayGameList(displayList.filter(e => e.name.toLowerCase().includes(value)));
});


async function init() {
    gameList = await loadGameList();
    displayGameList(gameList);
}

init();

